homeassistant/components/pi_hole/__init__.py

```diff
@@ -180,26 +108,88 @@ async def async_update_data():
         except HoleError as err:
             raise UpdateFailed(f"Failed to communicating with API: {err}")
 
+    entry.add_update_listener(async_update_listener)
+
     coordinator = DataUpdateCoordinator(
         hass,
         _LOGGER,
         name=name,
         update_method=async_update_data,
         update_interval=MIN_TIME_BETWEEN_UPDATES,
     )
-    hass.data[DOMAIN][name] = {
+    hass.data[DOMAIN][entry.entry_id] = {
         DATA_KEY_API: api,
         DATA_KEY_COORDINATOR: coordinator,
     }
 
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN)
-    )
+    for platform in PLATFORMS:
```

Comment:
Select here what platforms to load based on api key set or not.

Reply:
Moving the platform selection causes a problem. When the user add or remove API key from options, previously setup `switch` or `binary_sensor` cannot be properly unloaded without store the previous status somewhere.

Reply:
How can they not be unloaded? Please explain in more detail.

Reply:
For example, the user sets up the integration without an API key, so the binary sensor will be loaded. Then the user adds API key through option flow. During reload, `async_unload_entry` will try to unload `switch` instead of `binary_sensor`.

Reply:
Store the current platform names in hass.data in `async_setup_entry`.